File: app/src/androidTest/java/app/familygem/ExportTest.java
package app.familygem;

// Test di esportazione di un albero (con file media sparsi nello storage) verso GEDCOM e backup ZIP

import android.content.Context;
import android.net.Uri;
import android.os.Environment;
import androidx.documentfile.provider.DocumentFile;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;
import org.apache.commons.io.FileUtils;
import org.junit.Test;
import org.junit.runner.RunWith;
import java.io.File;
import java.io.InputStream;
import static org.junit.Assert.*;

@RunWith(AndroidJUnit4ClassRunner.class)
public class ExportTest {

	Context testContext; // Contesto del test (per accedere alle risorse in /assets)
	Context appContext; // Contesto di app.familylab

	@Test
	public void vaiCoiTest() throws Exception {
		testContext = InstrumentationRegistry.getInstrumentation().getContext();
		appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();

		// Eseguire questi 4 metodi uno alla volta:
		caricaFileGedcom(); // Dopodiché importa '/Gedcom/media.ged'
		//mettiFileMedia(); // Dopodiché aggiungi gli URI '/Uri' e '/SDCARD/Privata'
		//esportaGedcom();
		//esportaBackup();
	}

	// Carica /assets/media.ged in /GEDCOM/media.ged del device
	// Elimina l'albero 'media' esistente
	void caricaFileGedcom() throws Exception {
		InputStream inputStream = testContext.getAssets().open("media.ged");
		assertNotNull( inputStream );
		File gedcomFile = new File( Environment.getExternalStorageDirectory() + "/GEDCOM",
				"media.ged" );
		assertNotNull( gedcomFile );
		FileUtils.copyInputStreamToFile( inputStream, gedcomFile );
		assertTrue( gedcomFile.isFile() );

		for( Settings.Tree alb : Global.settings.trees ) {
			if( alb.title.equals("media") )
				Global.settings.deleteTree( alb.id );
		}
	}

	// Mette alcuni file in varie cartelle di Android
	// Aggiunge il path ai percorsi dell'albero
	void mettiFileMedia() throws Exception {
		Settings.Tree ultimoAlb = Global.settings.trees.get(Global.settings.trees.size() - 1);

		// PDF in external storage
		String percorso0 = appContext.getExternalFilesDir( String.valueOf(ultimoAlb.id) ).getPath();
		InputStream input = testContext.getAssets().open("È Carmelo.pdf");
		File external = new File( percorso0, "È Carmelo.pdf" );
		FileUtils.copyInputStreamToFile( input, external );
		assertTrue( external.isFile() );

		String percorso1 = Environment.getExternalStorageDirectory().getPath() + "/Percorso";
		File pathFile = new File( percorso1, "path.txt" );
		assertNotNull( pathFile );
		assertTrue( pathFile.canWrite() );
		FileUtils.writeStringToFile( pathFile, pathFile.getPath(), "UTF-8" );
		assertTrue( pathFile.isFile() );


		// Percorso di due txt omonimi in due sottocartelle
		String percorso2 = Environment.getExternalStorageDirectory().getPath() + "/Percorso Bis";

		// Txt da accedere con percorso, non riesce a scrivere in Android 10 e 11
		File pathFilePrimo = new File( percorso2 + "/primo", "omonimo.txt" );
		assertNotNull( pathFilePrimo );
		FileUtils.writeStringToFile( pathFilePrimo, pathFilePrimo.getPath(), "UTF-8" );
		assertTrue( pathFilePrimo.isFile() );

		// Altro file txt con lo stesso nome ma in un'altra sottocartella
		File pathFileSecondo = new File( percorso2 + "/secondo", "omonimo.txt" );
		assertNotNull( pathFileSecondo );
		FileUtils.writeStringToFile( pathFileSecondo, pathFileSecondo.getPath(), "UTF-8" );
		assertTrue( pathFileSecondo.isFile() );


		// Percorsi nelle preferenze
		assertEquals( ultimoAlb.title, "media" );
		//if( !ultimoAlb.cartelle.contains(percorso1) )
		ultimoAlb.dirs.add( percorso0 );
		assertTrue( ultimoAlb.dirs.contains(percorso0) );
		ultimoAlb.dirs.add( percorso1 );
		assertTrue( ultimoAlb.dirs.contains(percorso1) );
		ultimoAlb.dirs.add( percorso2 );
		assertTrue( ultimoAlb.dirs.contains(percorso2) );
		Global.settings.openTree = ultimoAlb.id;
		Global.settings.save();

		// File txt da prendere come Uri
		String percorso3 = Environment.getExternalStorageDirectory().getPath() + "/Uri";
		File uriFile = new File( percorso3, "uri.txt" );
		assertNotNull( uriFile );
		FileUtils.writeStringToFile( uriFile, DocumentFile.fromFile(uriFile).getUri().toString(), "UTF-8" );
		assertTrue( uriFile.isFile() );

		// Immagine nella scheda SD
		File[] luoghi = appContext.getExternalFilesDirs(null);
		for( File luogo : luoghi ) {
			if( !luogo.getPath().startsWith( Environment.getExternalStorageDirectory().getPath() ) ) {
				String sdFolder = luogo.getPath().substring(0, luogo.getPath().indexOf("Android/")) + "Privata";
				input = testContext.getAssets().open("anna.webp");
				assertNotNull( input );
				File file = new File( sdFolder, "anna.webp" );
				assertNotNull( file );
				FileUtils.copyInputStreamToFile( input, file );
				assertTrue( file.isFile() );
			}
		}
	}

	// Esporta in /Documents l'ultimo albero in 2 file: GEDCOM e ZIP coi media
	void esportaGedcom() {
		Settings.Tree ultimoAlb = Global.settings.trees.get(Global.settings.trees.size() - 1);
		//assertEquals( ultimoAlb.nome, "media" );
		int idAlbero = ultimoAlb.id;

		File documentsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS);
		if( !documentsDir.exists() ) documentsDir.mkdir();

		File fileGedcom = new File( documentsDir, "Küçük ağaç.ged" );
		Exporter esp = new Exporter( appContext );
		assertTrue( esp.openTree( idAlbero ) );
		assertNull( esp.successMessage);
		assertNull( esp.errorMessage);
		assertTrue( esp.exportGedcom(Uri.fromFile(fileGedcom)) );
		assertTrue( fileGedcom.isFile() );
		assertEquals( esp.successMessage, appContext.getString(R.string.gedcom_exported_ok) );
		s.l( esp.successMessage);

		File fileGedcomZip = new File( documentsDir, "ਸੰਕੁਚਿਤ.zip" );
		Exporter esp2 = new Exporter( appContext );
		assertTrue( esp2.openTree( idAlbero ) );
		boolean result = esp2.exportGedcomToZip(Uri.fromFile(fileGedcomZip));
		s.l( esp2.errorMessage);
		assertTrue( result );
		assertEquals( esp2.successMessage, appContext.getString(R.string.zip_exported_ok) );
		assertTrue( fileGedcomZip.isFile() );
		s.l( esp2.successMessage);
	}

	// Esporta in /Documents l'ultimo albero come backup ZIP
	void esportaBackup() {
		File documentsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS);
		if( !documentsDir.exists() ) documentsDir.mkdir();
		File fileBackup = new File( documentsDir, "Becàp olè.zip" );
		Exporter esp = new Exporter( appContext );
		Settings.Tree ultimoAlb = Global.settings.trees.get(Global.settings.trees.size()-1);
		assertTrue( esp.openTree( ultimoAlb.id ) );
		boolean result = esp.exportZipBackup( null, -1, Uri.fromFile(fileBackup) );
		s.l( esp.errorMessage);
		assertTrue( result );
		assertEquals( esp.successMessage, appContext.getString(R.string.zip_exported_ok) );
		assertTrue( fileBackup.isFile() );
		s.l( esp.successMessage);
	}
}

File: app/src/androidTest/java/app/familygem/ExportTest.kt
package app.familygem

import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Environment
import androidx.test.espresso.Espresso
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.ext.junit.rules.activityScenarioRule
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.GrantPermissionRule
import app.familygem.util.FileUtil
import app.familygem.util.TreeUtil
import kotlinx.coroutines.test.runTest
import org.apache.commons.io.FileUtils
import org.junit.Assert.*
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith
import java.io.File
import java.io.FileOutputStream

/**
 * Exports a tree with media files to GEDCOM and ZIP backup.
 */
@RunWith(AndroidJUnit4ClassRunner::class)
class ExportTest {

    @get:Rule
    val writePermissionRule: TestRule = if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.Q) {
        GrantPermissionRule.grant(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
    } else GrantPermissionRule.grant()

    @get:Rule
    val activityScenarioRule = activityScenarioRule<LauncherActivity>()

    private lateinit var testContext: Context // Test context (to access files in /assets)
    private lateinit var appContext: Context // Context of app.familygem

    @Before
    fun setup() {
        testContext = InstrumentationRegistry.getInstrumentation().context
        appContext = InstrumentationRegistry.getInstrumentation().targetContext
    }

    @Test
    fun exportTreeWithMediaFiles() = runTest {
        Global.settings.trees.clear()
        assertEquals(Global.settings.trees.size, 0)

        onView(withId(R.id.fab)).perform(click())
        onView(withId(R.id.new_import_gedcom)).check(matches(isDisplayed()))

        // Imports /assets/media.ged as a new tree
        val inputStream = testContext.assets.open("media.ged")
        assertNotNull(inputStream)
        val gedcomFile = File(appContext.cacheDir, "media.ged")
        assertNotNull(gedcomFile)
        FileOutputStream(gedcomFile).use { inputStream.copyTo(it) }
        assertTrue(gedcomFile.isFile)
        // This import fails on Android 4, 5 and 6 throwing a strange OutOfMemory error
        TreeUtil.importGedcom(appContext, Uri.fromFile(gedcomFile), {}, {})
        Espresso.pressBack()
        assertEquals(Global.settings.trees.size, 1)
        val mediaTree = Global.settings.trees[Global.settings.trees.size - 1]
        assertEquals(mediaTree.title, "media")

        // Puts some files in various Android folders

        // PDF in external storage
        val path1 = appContext.getExternalFilesDir(mediaTree.id.toString())!!.path
        val input = testContext.assets.open("È Carmelo.pdf")
        val pdfFile = File(path1, "È Carmelo.pdf")
        FileOutputStream(pdfFile).use { input.copyTo(it) }
        assertTrue(pdfFile.isFile)

        // Scoped Storage of Android 10 limits access only to certain folders
        val documentsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS)
        if (!documentsDir.exists()) documentsDir.mkdir() // Necessary
        assertTrue(documentsDir.exists())
        assertTrue(documentsDir.isDirectory)

        // TXT in shared storage
        val subFolder = "SubFolder"
        var path2 = File(Environment.getExternalStorageDirectory(), subFolder)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            path2 = File(documentsDir, subFolder)
        }
        val pathFile = File(path2, "path.txt")
        assertNotNull(pathFile)
        FileUtils.writeStringToFile(pathFile, pathFile.path, "UTF-8")
        assertTrue(pathFile.isFile)
        assertNotEquals(pathFile.length(), 0)

        // Two TXT files with same name in two subfolders
        val subFolderBis = "SubFolder Bis"
        var path3 = File(Environment.getExternalStorageDirectory(), subFolderBis)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            path3 = File(documentsDir, subFolderBis)
        }
        val homonym = "homonym.txt"
        val firstFile = File("$path3/first", homonym)
        assertNotNull(firstFile)
        FileUtils.writeStringToFile(firstFile, firstFile.path, "UTF-8")
        assertTrue(firstFile.isFile)
        assertNotEquals(firstFile.length(), 0)
        // Second TXT file with same name but in another subfolder
        val secondFile = File("$path3/second", homonym)
        assertNotNull(secondFile)
        FileUtils.writeStringToFile(secondFile, secondFile.path, "UTF-8")
        assertTrue(secondFile.isFile)
        assertNotEquals(secondFile.length(), 0)

        // Paths in preferences
        mediaTree.dirs.add(path1)
        assertTrue(mediaTree.dirs.contains(path1))
        mediaTree.dirs.add(path2.absolutePath)
        assertTrue(mediaTree.dirs.contains(path2.absolutePath))
        mediaTree.dirs.add(path3.absolutePath)
        assertTrue(mediaTree.dirs.contains(path3.absolutePath))
        Global.settings.save()

        // TODO: test tree Uri

        // Exports in /Documents the tree as GEDCOM file
        val treeId = mediaTree.id
        assertEquals(treeId, 1)
        val fileGedcom = File(documentsDir, "Küçük ağaç.ged")
        val exporter1 = Exporter(appContext)
        assertTrue(exporter1.openTree(treeId))
        assertNull(exporter1.successMessage)
        assertNull(exporter1.errorMessage)
        assertTrue(exporter1.exportGedcom(Uri.fromFile(fileGedcom)))
        assertTrue(fileGedcom.isFile)
        assertNotEquals(fileGedcom.length(), 0)
        assertEquals(exporter1.successMessage, appContext.getString(R.string.gedcom_exported_ok))

        // Exports in /Documents the tree as zipped GEDCOM with media files
        val fileGedcomZip = File(documentsDir, "ਸੰਕੁਚਿਤ.zip")
        val exporter2 = Exporter(appContext)
        assertTrue(exporter2.openTree(treeId))
        val result1 = exporter2.exportGedcomToZip(Uri.fromFile(fileGedcomZip))
        assertTrue(result1)
        assertEquals(exporter2.successMessage, appContext.getString(R.string.zip_exported_ok))
        assertTrue(fileGedcomZip.isFile)

        // Exports in /Documents the tree as ZIP backup
        val fileBackup = File(documentsDir, "バックアップ.zip")
        val exporter3 = Exporter(appContext)
        assertTrue(exporter3.openTree(treeId))
        val result2 = exporter3.exportZipBackup(null, -1, Uri.fromFile(fileBackup))
        assertTrue(result2)
        assertEquals(exporter3.successMessage, appContext.getString(R.string.zip_exported_ok))
        assertTrue(fileBackup.isFile)

        // Deletes created files
        FileUtil.deleteFilesAndDirs(path2) // SubFolder
        assertFalse(pathFile.exists())
        assertFalse(path2.exists())
        FileUtil.deleteFilesAndDirs(path3) // SubFolder Bis
        assertFalse(firstFile.exists())
        assertFalse(secondFile.exists())
        assertFalse(path3.exists())

        // The exported zipped GEDCOM and ZIP backup should have 3 files each inside the /media folder
    }
}
